### ISSUE TITLE
[R] A shiny app to demonstrate image prediction from a pre-trained model

### ISSUE THREAD

**Issue by hetong007** [dmlc/mxnet#629]
[Shiny](http://shiny.rstudio.com/) is an R package which enables users to build web app with R as the computational backend. We can make an app demonstrating the image prediction to users in real time. 

One similar shiny app example is demonstrating image compression by svd: https://yihui.shinyapps.io/imgsvd/. For more examples of shiny, please visit http://shiny.rstudio.com/gallery/.

This issue is used to track the progress on this shiny app.

**Comment by thirdwing** [comment]
A very simple version has been done.

https://github.com/thirdwing/mxnet_shiny

**Comment by tqchen** [comment]
@thirdwing can we find anyplace to host it?

**Comment by hetong007** [comment]
This might be a good place: https://www.shinyapps.io/

**Comment by thirdwing** [comment]
At least you should have a look at http://107.170.74.195:3838/mxnet/

Then we can talk about deployment in other places.

**Comment by thirdwing** [comment]
It is a weak server with only 1GB memory, so it is really slow.

**Comment by tqchen** [comment]
Looks nice, would be great if user can also simply provide an URL to the image on the web

**Comment by tqchen** [comment]
cc @mli

**Comment by mli** [comment]
let's do the following things

1. put things into docker to easy deploy and upgrade
2. find a proper aws instance you want, say c4.2x large, i will create one (or a few if there are a lot of traffic)
3. ping @phunterlau to redirect a sub-domain of dmlc.io or dmlc.ml to that ip address

**Comment by thirdwing** [comment]
OK. I will make a docker for this.

**Comment by phunterlau** [comment]
got it, what subdomain do you want? I can simply add a cname to it. `shiny.dmlc.ml` for example?

**Comment by phunterlau** [comment]
by the way, I see you use port number with path. If possible, with the AWS server, can you get a standalone IP, e.g. 107.170.74.195 so I can add an 'A' record to `shiny.dmlc.ml` to this IP. It is the best and universal solution, otherwise I may need to add http redirect which may have issue of multiple backend machines (multiple A record is very fine).

**Comment by thirdwing** [comment]
OK.

@phunterlau Now the shiny app is at http://107.170.74.195/

@tqchen It can take url.

**Comment by phunterlau** [comment]
nice, here we go <http://shiny.dmlc.ml/> if you want to change to some other sub-domain, feel free to ask here. Just an easy A record.

**Comment by phunterlau** [comment]
a small problem while using URL: if give some URL images for example https://i.ytimg.com/vi/tntOCGkgt98/maxresdefault.jpg it shows an error `Attempted to assign value to a read-only reactivevalues object` I tried multiple URL links, none worked.

**Comment by hetong007** [comment]
I successfully used the URL about 30 mins ago, but I got the same error now.

**Comment by thirdwing** [comment]
Please wait for a while. I am updating it.

**Comment by thirdwing** [comment]
It should work now. Really inexperienced in shiny.

**Comment by hetong007** [comment]
Thanks, it works now, but it seems to have too much workload for the server. I got the error message `cannot popen '/usr/bin/which 'convert' 2>/dev/null', probable reason 'Cannot allocate memory'`

**Comment by thirdwing** [comment]
`convert` is used by `imager` package and it is a little expensive for a server with 1 GB RAM, especially dealing with large figures.

**Comment by tqchen** [comment]
@thirdwing @phunterlau  @mli  We will need to add one more functionality. Allow user to give feedback to rate the result, and provide a text box of label that they think are correct, this can help us to collect some data to improve the model

**Comment by phunterlau** [comment]
@tqchen good idea, we can have a more powerful machine for more users, and some buttons on the page of collecting feedbacks (e.g. giving each answer a "like" button etc). We can post it to weibo and reddit when ready.

P.S. I have also setup <http://shiny.mxnt.ml/> for this demo, since it is mxnet project. (I have registered many DMLC domains, try mxnt.ml xgboost.ml :-) )

**Comment by thirdwing** [comment]
BTW, personally I don't think shiny is designed for such a purpose, we
might need to switch to a more powerful web framework.

Let me talk with the authors of shiny and get back to you.

On Sat, Nov 21, 2015 at 2:08 PM, Hongliang Liu <notifications@github.com>
wrote:

> @tqchen <https://github.com/tqchen> good idea, we can have a more
> powerful machine for more users, and some buttons on the page of collecting
> feedbacks (e.g. giving each answer a "like" button etc). We can post it to
> weibo and reddit when ready.
>
> P.S. I have also setup http://shiny.mxnt.ml/ for this demo, since it is
> mxnet project. (I have registered many DMLC domains, try mxnt.ml
> xgboost.ml :-) )
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmlc/mxnet/issues/629#issuecomment-158672916>.
>



-- 
Qiang Kou
qkou@umail.iu.edu
School of Informatics and Computing, Indiana University

**Comment by hetong007** [comment]
By using shiny we emphasize the power of **MXNetR**. A more powerful web framework might be more efficient, but not that attractive for the R's community. 

If you mean the computational tasks are too heavy tor a shiny app and you want to discuss it with shiny authors, that would be great :)

**Comment by thirdwing** [comment]
The limitation is in the free version of shiny server. As I know, using free version, there will only one R process no matter how people are using this app. So people might queue to see the results.

I am not very sure about this. Let me double check with Rstudio people.

KK

**Comment by mli** [comment]
the single process limit will be big problem then...

it seems that we don't the fancy interactive feature provided by shiny.. so we can use some other high performance framework then. one option is using go.

**Comment by futurely** [comment]
If handling more user traffic is a concern, the architecture of the app needs to be [upgraded to be distributed asynchronous pipeline](https://aws.amazon.com/architecture). A similar not exactly the same scenario is [media sharing](http://media.amazonwebservices.com/architecturecenter/AWS_ac_ra_mediasharing_09.pdf) in which media files are first uploaded into S3 and processed, e.g. resized and classified, by dedicated server instance. The user facing front-end web server only handles the web requests and returns the corresponding responses when the classification results are notified by the back-end heavy lifting servers. The benefit of decoupling IO instensive web hosting and CPU/GPU instensive computation is more flexible choice of server types, dynamic scaling and non-interrupted web service in case of disaster (think corrupted or huge images) recovery.

**Comment by tqchen** [comment]
closing as shiny app already pushed in, looking forward to the blogpost
